main.tf

```diff
@@ -14,100 +14,162 @@
-  ip_version = "${var.ip_version}"
+  ip_version = var.ip_version
 }
 
 # HTTP proxy when ssl is false
 resource "google_compute_target_http_proxy" "default" {
-  project = "${var.project}"
-  count   = "${var.http_forward ? 1 : 0}"
+  project = var.project
+  count   = var.http_forward ? 1 : 0
   name    = "${var.name}-http-proxy"
-  url_map = "${element(compact(concat(list(var.url_map), google_compute_url_map.default.*.self_link)), 0)}"
+  url_map = compact(
+  concat([
+    var.url_map], google_compute_url_map.default.*.self_link),
+  )[0]
 }
 
 # HTTPS proxy  when ssl is true
 resource "google_compute_target_https_proxy" "default" {
-  project          = "${var.project}"
-  count            = "${var.ssl ? 1 : 0}"
+  project          = var.project
+  count            = var.ssl ? 1 : 0
   name             = "${var.name}-https-proxy"
-  url_map          = "${element(compact(concat(list(var.url_map), google_compute_url_map.default.*.self_link)), 0)}"
-  ssl_certificates = ["${compact(concat(var.ssl_certificates, google_compute_ssl_certificate.default.*.self_link))}"]
+  url_map          = compact(
+  concat([
+    var.url_map], google_compute_url_map.default.*.self_link), )[0]
+  ssl_certificates = compact(concat(var.ssl_certificates, google_compute_ssl_certificate.default.*.self_link, ), )
 }
 
 resource "google_compute_ssl_certificate" "default" {
-  project     = "${var.project}"
-  count       = "${(var.ssl && !var.use_ssl_certificates) ? 1 : 0}"
+  project     = var.project
+  count       = var.ssl && false == var.use_ssl_certificates ? 1 : 0
   name_prefix = "${var.name}-certificate-"
-  private_key = "${var.private_key}"
-  certificate = "${var.certificate}"
+  private_key = var.private_key
+  certificate = var.certificate
 
-  lifecycle = {
+  lifecycle {
     create_before_destroy = true
   }
 }
 
 resource "google_compute_url_map" "default" {
-  project         = "${var.project}"
-  count           = "${var.create_url_map ? 1 : 0}"
+  project         = var.project
+  count           = var.create_url_map ? 1 : 0
   name            = "${var.name}-url-map"
-  default_service = "${google_compute_backend_service.default.0.self_link}"
+  default_service = google_compute_backend_service.default[keys(var.backends)[0]].self_link
 }
 
 resource "google_compute_backend_service" "default" {
-  project         = "${var.project}"
-  count           = "${length(var.backend_params)}"
-  name            = "${var.name}-backend-${count.index}"
-  port_name       = "${element(split(",", element(var.backend_params, count.index)), 1)}"
-  protocol        = "${var.backend_protocol}"
-  timeout_sec     = "${element(split(",", element(var.backend_params, count.index)), 3)}"
-  backend         = ["${var.backends["${count.index}"]}"]
-  health_checks   = ["${element(google_compute_http_health_check.default.*.self_link, count.index)}"]
-  security_policy = "${var.security_policy}"
-  enable_cdn      = "${var.cdn}"
+  for_each = toset(keys(var.backends))
```

Comment:
You've changed logic, instead of length of packend_params you use backends

Reply:
They should match. I'd combine them in one var object.